haskellPackages: allow hnix-store-{core,remote} 0.5

Anton-Latukha opened the pull request:
These were blocked due to mutual desigion during me<>`sternenseemann`
discussion.

In short:

I shipped my own work (to support GHC 9.0) in the 0.5 releases of
`hnix-store-{core,remote}`.

These packages are really used only by `hnix` itself, and instead of maintaining
them in Nixpkgs & reacting on `hnix` release, we decided to hold them back &
switch to these versions when `hnix` provides support for them.

I just (today) released `hnix` 0.14 & it requires `hnix-store-{core,remote}`
0.5.

If you would look at dependency tree:
https://packdeps.haskellers.com/reverse/hnix-store-core
https://packdeps.haskellers.com/reverse/hnix-store-remote

It shows that `hnix` currently is the only alive use of these projects.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Anton-Latukha committed:
haskellPackages: allow hnix-store-{core,remote} 0.5

These were blocked due to mutual desigion during me<>`sternenseemann`
discussion.
https://github.com/haskell-nix/hnix/issues/952

In short:

I shipped my own work (to support GHC 9.0) in the 0.5 releases of
`hnix-store-{core,remote}`.

These packages are really used only by `hnix` itself, and instead of maintaining
them in Nixpkgs & reacting on `hnix` release, we decided to hold them back &
switch to these versions when `hnix` provides support for them.

I just (today) released `hnix` 0.14 & it requires `hnix-store-{core,remote}`
0.5.

If you would look at dependency tree:
https://packdeps.haskellers.com/reverse/hnix-store-core
https://packdeps.haskellers.com/reverse/hnix-store-remote

It shows that `hnix` currently is the only alive use of these projects.

Anton-Latukha committed:
haskellPackages: add myself as a maintainer for packages

sternenseemann commented:
If you run `./maintainers/scripts/haskell/regenerate-hackage-packages.sh` you should be able to see the effects of your changes and test them as well (it generates `pkgs/development/haskell-modules/hackage-packages.nix`).

Anton-Latukha commented:
Recieved a clean diff:
```diff
diff --git a/pkgs/development/haskell-modules/hackage-packages.nix b/pkgs/development/haskell-modules/hackage-packages.nix
index 364581829a7..a00deb16fff 100644
--- a/pkgs/development/haskell-modules/hackage-packages.nix
+++ b/pkgs/development/haskell-modules/hackage-packages.nix
@@ -131462,41 +131462,10 @@ self: {
        ];
        description = "Haskell implementation of the Nix language";
        license = lib.licenses.bsd3;
+       maintainers = with lib.maintainers; [ Anton-Latukha ];
      }) {};
 
   "hnix-store-core" = callPackage
-    ({ mkDerivation, algebraic-graphs, attoparsec, base
-     , base16-bytestring, base64-bytestring, binary, bytestring, cereal
-     , containers, cryptohash-md5, cryptohash-sha1, cryptohash-sha256
-     , cryptohash-sha512, directory, filepath, hashable, hspec
-     , lifted-base, monad-control, mtl, nix-derivation, process, saltine
-     , tasty, tasty-discover, tasty-golden, tasty-hspec, tasty-hunit
-     , tasty-quickcheck, temporary, text, time, unix
-     , unordered-containers, vector
-     }:
-     mkDerivation {
-       pname = "hnix-store-core";
-       version = "0.4.3.0";
-       sha256 = "17l7198s3dzjx416nplinyghbqgignvpvxkww241gjv4p23izf77";
-       libraryHaskellDepends = [
-         algebraic-graphs attoparsec base base16-bytestring
-         base64-bytestring bytestring cereal containers cryptohash-md5
-         cryptohash-sha1 cryptohash-sha256 cryptohash-sha512 directory
-         filepath hashable lifted-base monad-control mtl nix-derivation
-         saltine text time unix unordered-containers vector
-       ];
-       testHaskellDepends = [
-         attoparsec base base16-bytestring base64-bytestring binary
-         bytestring containers directory filepath hspec process tasty
-         tasty-golden tasty-hspec tasty-hunit tasty-quickcheck temporary
-         text unix
-       ];
-       testToolDepends = [ tasty-discover ];
-       description = "Core effects for interacting with the Nix store";
-       license = lib.licenses.asl20;
-     }) {};
-
-  "hnix-store-core_0_5_0_0" = callPackage
     ({ mkDerivation, algebraic-graphs, attoparsec, base
      , base16-bytestring, base64-bytestring, binary, bytestring, cereal
      , containers, cryptonite, directory, filepath, hashable, hspec
@@ -131526,27 +131495,10 @@ self: {
        testToolDepends = [ tasty-discover ];
        description = "Core effects for interacting with the Nix store";
        license = lib.licenses.asl20;
-       hydraPlatforms = lib.platforms.none;
+       maintainers = with lib.maintainers; [ Anton-Latukha ];
      }) {};
 
   "hnix-store-remote" = callPackage
-    ({ mkDerivation, attoparsec, base, binary, bytestring, containers
-     , hnix-store-core, mtl, network, nix-derivation, text, time
-     , unordered-containers
-     }:
-     mkDerivation {
-       pname = "hnix-store-remote";
-       version = "0.4.3.1";
-       sha256 = "1wp6z5f13fkgh3mdzr1zcqdvsx1j90l5isf671gsg43brsv4c2vl";
-       libraryHaskellDepends = [
-         attoparsec base binary bytestring containers hnix-store-core mtl
-         network nix-derivation text time unordered-containers
-       ];
-       description = "Remote hnix store";
-       license = lib.licenses.asl20;
-     }) {};
-
-  "hnix-store-remote_0_5_0_0" = callPackage
     ({ mkDerivation, attoparsec, base, binary, bytestring, containers
      , cryptonite, hnix-store-core, mtl, network, nix-derivation, text
      , time, unordered-containers
@@ -131562,7 +131514,7 @@ self: {
        ];
        description = "Remote hnix store";
        license = lib.licenses.asl20;
-       hydraPlatforms = lib.platforms.none;
+       maintainers = with lib.maintainers; [ Anton-Latukha ];
      }) {};
 
   "hnn" = callPackage
```

sternenseemann commented:
Feel free to commit this as well here (or right before unmarking as WIP to avoid merge conflicts).

Anton-Latukha committed:
haskellPackages: regenerated hackage-packages.nix

Anton-Latukha commented:
Just an idea: if to produce a lot of files (maybe a file per project/package) - there probably would be drastically fewer regeneration diff conflicts (I think).

sternenseemann commented:
@ofborg build haskellPackages.hnix

sternenseemann commented:
When was the hnix release supporting the store packages?

maralorn commented:
> Just an idea: if to produce a lot of files (maybe a file per project/package) - there probably would be drastically fewer regeneration diff conflicts (I think).

I don‘t think this is necessary. Merge conflicts in the hackage-packages.nix are not that common in the first place.
I have never noticed a merge conflict caused by two unrelated packages. I assume the 99% of all our merge conflicts come from changes to the same package. And those are honest, real merge conflicts.

Anton-Latukha commented:
> When was the hnix release supporting the store packages?

Ttoday. After releases, if I can I contribute patches to Nixpkgs.

Gabriel439 commented:
I believe this is safe to merge into `haskell-updates` as the latest version of `hnix` on Hackage supports the latest versions of `hnix-store-{core,remote}`.  I can merge this myself tomorrow if nobody objects

sternenseemann commented:
I guess we'll need to decide what to do about our hnix patches as well. The commits I originally reverted no longer apply with 0.14.*, so we either need update these patches and vendor them in or bite the bullet and override `relude_1_0_0_1` and `semialign_1_2`, but I have no idea if this causes any issues in reverse dependencies…

sternenseemann committed:
haskellPackages.hnix: build with relude 1.0.0.1 and semialign 1.2

hnix needs these versions since 0.13, but we previously patched it to
use the versions tracked in stackage because that reduces the risk of
multiple versions of a package being propagated in the dependency tree
and breaking a build.

One major release later, patching hnix has become quite cumbersome, so
we'll bite the bullet for now and return to this approach if any
problems come up.

sternenseemann committed:
haskellPackages.nix-linter: unbreak for hnix 0.14

Change to pname is necessary in order to not confuse jailbreak-cabal.

sternenseemann commented:
@ofborg build haskellPackages.nix-linter haskellPackages.dhall-nix haskellPackages.dhall-nixpkgs haskellPackages.yarn2nix

(`update-nix-fetchgit` builds locally, but its test suite fails, I guess that is something @expipiplus1 can have a look at later.)
